airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_incremental.py

```diff
@@ -189,17 +222,36 @@ def test_read_sequential_slices(
         # We sometimes have duplicate identical state messages in a stream which we can filter out to speed things up
         checkpoint_messages = [message for index, message in enumerate(checkpoint_messages) if message not in checkpoint_messages[:index]]
 
-        # To avoid spamming APIs we only test a fraction of slices
-        num_slices_to_test = 1 if len(checkpoint_messages) <= 5 else len(checkpoint_messages) // 5
-        for message_batch in checkpoint_messages[::num_slices_to_test]:
+        # To avoid spamming APIs we only test a fraction of batches (10%) and enforce a minimum of 10 tested
+        min_batches_to_test = 10
+        sample_rate = len(checkpoint_messages) // min_batches_to_test
+        stream_name_to_per_stream_state = dict()
+        for idx, message_batch in enumerate(checkpoint_messages):
             assert len(message_batch) > 0 and message_batch[0].type == Type.STATE
-            current_state = message_batch[0]
-            output = docker_runner.call_read_with_state(connector_config, configured_catalog_for_incremental, current_state.state.data)
+
+            if is_per_stream:
+                # Including all the latest state values from previous batches, update the combined stream state
+                # with the current batch's stream state and then use it in the following read() request
+                current_state = message_batch[0].state
+                if current_state and current_state.type == AirbyteStateType.STREAM and hasattr(current_state, "stream"):
+                    per_stream = current_state.stream
+                    if per_stream.stream_state:
+                        stream_name_to_per_stream_state[per_stream.stream_descriptor.name] = (
+                            per_stream.stream_state.dict() if per_stream.stream_state else {}
+                        )
+                state_input = [latest_per_stream for latest_per_stream in stream_name_to_per_stream_state.values()]
+                complete_state = stream_name_to_per_stream_state
+            else:
+                state_input = message_batch[0].state.data
+                complete_state = message_batch[0].state.data
+
+            if len(checkpoint_messages) >= min_batches_to_test and idx % sample_rate != 0:
```

Comment:
we used to sample records by deriving how much to increment before the loop. However, since we still need to maintain the state of all streams up to this point, we still perform the updates for each batch. But then we skip over the read and compare checks for most batches to not slow down the tests too much

Reply:
I'm concerned this change can lead to different test outputs for the existing connectors.

Reply:
I'm curious in what way would this introduce a testing regression? The way the test is designed, increasing the granularity would mean more accuracy that batches are correct, if they weren't then it would mean. we caught a bug.

Or if you meant literal test output, we're not outputting any data from these tests, we just perform assertions